src/node/pipeline.spec.ts

```diff
@@ -0,0 +1,116 @@
+// node/pipeline.spec.ts
+
+/*
+ * Copyright (c) 2021 Check Digit, LLC
+ *
+ * This code is licensed under the MIT license (see LICENSE.txt for details).
+ */
+
+/* eslint-disable deprecate/function */
+
+import assert from 'assert';
+import { Readable, Writable } from 'stream';
+
+import { toString } from '../sink';
+
+import pipeline from './pipeline';
+
+async function* passThru<T>(iterable: AsyncIterable<T>): AsyncGenerator<T> {
+  for await (const thing of iterable) {
+    yield thing;
+  }
+}
+
+async function validateReadable(stream: Readable, expected: string) {
+  assert.ok(stream.readable);
+
+  let ended = false;
+  stream.on('end', () => {
+    ended = true;
+  });
+
+  let finished = false;
+  stream.on('finish', () => {
+    finished = true;
+  });
+
+  assert.strictEqual(await toString(stream), expected);
+
+  assert.ok(finished);
+  assert.ok(ended);
+
+  assert.strictEqual(await toString(stream), '');
+}
+
+describe('pipeline', () => {
+  it('throws error if source is a Buffer object', async () => {
+    // tracking this behavior.  I don't believe this should crash, possibly a bug in node stream.pipeline implementation.
+    await assert.rejects(async () => pipeline(Buffer.from('crash'), passThru, toString));
+  });
+
+  it('returns promise if last parameter is an async function', async () => {
+    const result1 = pipeline([undefined, 1, null, 2, true, 3, [4, [5], 6, 7]], passThru, toString);
+    const result2 = pipeline(Buffer.from('hello').values(), toString);
+    const result3 = pipeline('hello', toString);
+    assert.ok(typeof result1.then === 'function');
+    assert.ok(typeof result2.then === 'function');
+    assert.ok(typeof result3.then === 'function');
+    assert.strictEqual(await result1, 'undefined1null2true34,5,6,7');
+    assert.strictEqual(await result2, '104101108108111');
+    assert.strictEqual(await result3, 'hello');
+  });
+
+  it('returns ReadWriteStream if last parameter is an async generator', async () => {
+    await validateReadable(pipeline([undefined, 1, 2, true, 3], passThru), '12true3');
```

Comment:
Not quite following what causes the result here to be '12true3' instead of 'undefined12true3' as was the case in the 'result1' case in the test above?

Reply:
This is because anything involving a stream will swallow undefined and error on nulls.  But even though stream.pipeline allows this, maybe we should enforce the same behavior with non-stream pipelines.  I can't think of any valid reason to allow it, will create an issue to track.